Improve LSP17-ContractExtension explanation and move to Review (#282)
lukso-network/LIPs: LSPs/LSP-17-ContractExtension.md

## Abstract

> This proposal introduces two contract types:

Reviewer: Hmmm I would have said `types of contract`, as it makes it sound like Typescript type. But not a big deal

Smart contracts that adhere to the LSP17Extendable standard MUST include the `supportsInterface(..)` function, as specified in the [ERC165] standard, and MUST support the LSP17Extendable interfaceId.

### Behavior

#### Storing Extensions

> The standard offers flexibility in how extensions are mapped and stored. It does not enforce a specific method for mapping function selectors to the addresses of extension contracts, nor does it require a specific method for setting and retrieving these addresses. This flexibility allows developers to choose the most suitable approach for their specific use case.

Reviewer: Suggest to keep the wording generic first to describe the concept, instead of using the word `mapping` which might be more a programming language term

```suggestion
The standard offers flexibility in how extensions are linked and stored. It does not enforce a specific method for linking function selectors to the addresses of extension contracts, nor does it require a specific method for setting and retrieving these addresses. This flexibility allows developers to choose the most suitable approach for their use case.
```

**Example of a Simple Mapping Approach**

> A straightforward way to implement this mapping is through a simple mapping data type, and a simple function to set the address of the extension to the function selector. For instance:

Reviewer: Suggest adding language name for the example (Since standards should be language agnostics)

```suggestion
A straightforward way to implement this mapping is through a simple mapping data type, and a simple function to set the address of the extension to the function selector. For instance in Solidity:
```

> A straightforward way to implement this mapping is through a simple mapping data type, and a simple function to set the address of the extension to the function selector. For instance:

Reviewer: ```suggestion
A straightforward way to implement this linking is through a simple mapping data type, and a simple function to set the address of an extension contract for a non-native function selector. For instance:
```

```solidity
mapping (bytes4 => address) extensionStorage;

function setExtension(bytes4 functionSelector, address extension) {
  extensionStorage[functionSelector] = extension;
}
```

**Storing Extensions in ERC725Y Storage**

> Contracts that conform to the ERC725Y standard COULD use an alternative approach to store extension information in the ERC725Y storage. The ERC725Y storage is a mapping from bytes32 to bytes, and the extension information can be structured as follows:

Reviewer: Should these be quoted as code maybe?

```suggestion
Contracts that conform to the ERC725Y standard COULD use an alternative approach to store extension information in the ERC725Y storage. The ERC725Y storage is a mapping from `bytes32` to `bytes`, and the extension information can be structured as follows:
```

```json
{
  "name": "LSP17Extension:<bytes4>",
  "key": "0xcee78b4094da860110960000<bytes4>",
  "keyType": "Mapping",
  "valueType": "address",
  "valueContent": "Address"
}
```

> In this structure, <bytes4> represents the function selector of the contract. For a specific function selector, such as 0xaabbccdd, the data key would be constructed as follows:

Reviewer: Suggested rewording

```suggestion
In this structure, <bytes4> represents the selector of the function to add to the contract. For a specific function selector, such as 0xaabbccdd, the data key would be constructed as follows:
```

```js
0xcee78b4094da860110960000aabbccdd00000000000000000000000000000000;
```

> To add an extension, users can call `setData(bytes32 dataKey, bytes dataValue)` with the data key representing the function selector and the data value being the address of the extension.

Reviewer: ```suggestion
To add an extension, users can call `setData(bytes32 dataKey, bytes dataValue)` with the data key representing the function selector and the data value being the address of the extension contract.
```

```js
dataKey: 0xcee78b4094da860110960000aabbccdd00000000000000000000000000000000

> dataValue: [address of the extension]

Reviewer: ```suggestion
dataValue: [address of the extension contract]
```

```

> This approach leverages the ERC725Y standard's key-value storage mechanism to maintain a mapping of function selectors to extension addresses, providing a standardized and interoperable method for managing extensions in contracts that support ERC725Y.

Reviewer: ```suggestion
This approach leverages the ERC725Y standard's data key-value store to maintain a mapping of function selectors to extension addresses, providing a standardized and interoperable way to manage extensions in contracts that support ERC725Y.
```

#### Calling Extensions

Whenever a function is called on an extendable contract and the function does not exist, the fallback function of the extendable contract MUST call the function on the extension mapped using the `CALL` opcode.

The calldata MUST be appended with 52 extra bytes as follows:

- The `msg.sender` calling the extendable contract without any pad, MUST be 20 bytes.
- The `msg.value` received to the extendable contract, MUST be uint256.

> In case where the function selector does not correspond to a mapped extension within the extendable contract, the call SHOULD revert following the typical response observed when an a non existing function is called on a contract.

Reviewer: ```suggestion
In the case where the function selector does not correspond to a mapped extension within the extendable contract, the call SHOULD revert following the typical response observed when a non-existing function is called on a contract.
```

}
```

In this data structure:

<bytes4> represents the function selector called on the account contract. For instance, for the selector `0xaabbccdd`, the data key would be:

```js
0xcee78b4094da860110960000aabbccdd00000000000000000000000000000000;
```

> The boolean at the end of the data structure is utilized to indicate whether the value received by the extendable contract should be forwarded to the extension or remain within the extendable contract.

Reviewer: ```suggestion
The boolean at the end of the data structure is used to indicate whether the value received by the extendable contract should be forwarded to the extension or remain within the extendable contract.
```

> For example, if this boolean is set to true (represented as a `0x01` hex byte), it indicates that the value received should be forwarded to the extension:

Reviewer: ```suggestion
For example, if this boolean is set to `true` (represented as a `0x01` hex byte), it indicates that the value received should be forwarded to the extension:
```

> See [LSP2-ERC725YJSONSchema] to learn how to encode the data key, and the [**Mapping**](./LSP-2-ERC725YJSONSchema.md#mapping) section to learn the padding rules.

#### Supporting InterfaceIds

> In extendable contracts that implement the ERC165 standard, the `supportsInterface` function should be designed to additionally check for extended interfaceIds. This means that beyond the contract's native supported interfaceId, the function should also check support of extended interfaceIds.

Reviewer: ```suggestion
In extendable contracts that implement the ERC165 standard, the `supportsInterface` function should be designed to additionally check for extended ERC165 interface Ids. This means that beyond the contract's natively supported interface Id, the function should also check the support of extended interfaces.
```

> The `supportsInterface` function should call the `supportsInterface` function on the extension mapped to its function selector and return true or false based on the result.

Reviewer: > should call the `supportsInterface` function on the extension mapped to its function selector

are we talking about the extension mapped for the `supportsInterface(...)` function selector? If yes, maybe precise `the "supportsInterface" function selector (+ the bytes4 selector alongside to make it even easier for the reader :) )`.

Also suggest as usual marking `true` and `false` in code.

```suggestion
The `supportsInterface` function should call the `supportsInterface` function on the extension mapped to its function selector and return `true` or `false` based on the result.
```
